lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -35,82 +35,35 @@ namespace mlir::tt::ttnn {
-                                              targetTensorMemoryLayout.value())
-          : nullptr;
-
-  // Create the output memory config attribute.
-  ttnn::MemoryConfigAttr outputMemConfigAttr = ttnn::MemoryConfigAttr::get(
-      rewriter.getContext(),
-      ttnn::BufferTypeAttr::get(rewriter.getContext(), targetTensorBufferType),
-      ttnn::ShardSpecAttr::get(
-          op.getContext(),
-          ttnn::ShapeAttr::get(rewriter.getContext(),
-                               inputLayoutAttr.getMemref().getShape())),
-      outputMemLayoutAttr);
+// If the layout of the output result has changed as a result of applying a
+// workaround, this method transforms the layout back to the previous state
+// by inserting a ToLayoutOp after the op result output in order to maintain
+// the workarounds changes locally.
+//
+static void revertOutputLayout(wa::TTNNWorkaroundInterface &op,
+                               PatternRewriter &rewriter,
+                               wa::WorkaroundResults &workaroundResults,
+                               OpResult newOpResult) {
+  // Check if the data type of the output result has changed.
+  if (!workaroundResults.isModified()) {
+    return;
+  }
 
-  // Get the input operand type.
-  RankedTensorType inputOperandType =
-      mlir::cast<RankedTensorType>(inputOperand.get().getType());
-
-  // Create a ToLayoutOp to convert the input operand to the desired
-  // tensor layout, buffer type and memory layout.
-  return rewriter
-      .create<ttnn::ToLayoutOp>(
-          op.getLoc(),
-          ttnn::utils::createRankedTensorTypeWithEncoding(
-              inputOperandType,
-              inputLayoutAttr
-                  .withElementType(rewriter.getContext(), elementType)
-                  .withBufferType(rewriter.getContext(), targetTensorBufferType)
-                  .withMemoryLayout(rewriter.getContext(),
-                                    outputMemLayoutAttr)),
-          inputOperand.get(),
-          LayoutAttr::get(rewriter.getContext(), targetTensorLayout),
-          DataTypeAttr::get(rewriter.getContext(),
-                            inputLayoutAttr.getDataType()),
-          outputMemConfigAttr,
-          (targetTensorBufferType == ttnn::BufferType::SystemMemory)
-              ? nullptr
-              : utils::getOrInsertDevice(rewriter, op))
-      ->getResult(0);
+  // Insert the toLayoutOp after the op output.
+  rewriter.setInsertionPointAfter(op);
+
+  // Cast the data type back to the previous data type by inserting ToLayoutOp.
+  mlir::Value castLayoutOp = utils::createToLayoutOp(
+      op.getOperation(), newOpResult, rewriter,
+      workaroundResults.tensorLayoutResult.previousValue,
+      workaroundResults.tensorBufferTypeResult.previousValue,
+      workaroundResults.tensorMemoryLayoutResult.previousValue);
+
+  // Replace the new output result with the casted output result.
+  rewriter.replaceUsesWithIf(
```

Comment:
Just curious, did you try with `replaceAllOpUsesWith` and did it fail? I would assume it would be noop if we tried to replace `Value` with itself.

Reply:
I tried it, but it started behaving wrong in the sense that it started messing up with the op positon in IR and producing an error:
```
test/ttmlir/Dialect/TTNN/pooling/simple_maxpool2d.mlir:6:10: error: operand #0 does not dominate this use
    %1 = "ttir.max_pool2d"(%arg0, %0) <{kernel_height=2: si32, kernel_width=2: si32, stride_height=2: si32, stride_width=2: si32, dilation_height=1: si32, dilation_width=1: si32, ceil_mode=false, padding_left=0: si32, padding_right=0: si32, padding_top=0: si32, padding_bottom=0: si32}> : (tensor<1x128x128x32xbf16>, tensor<1x64x64x32xbf16>) -> tensor<1x64x64x32xbf16>
```